core/src/main/java/kafka/server/share/SharePartition.java

```diff
@@ -1343,21 +1346,31 @@ public boolean maybeAcquireFetchLock() {
     }
 
     /**
-     * Release the fetch lock once the records are fetched from the leader.
+     * Release the fetch lock once the records are fetched from the leader. It is imperative that the DelayedShareFetch instance
+     * that acquired the fetch lock should be the one releasing it.
+     * @param fetchId - The DelayedShareFetch instance uuid that is trying to release the fetch lock.
      */
-    void releaseFetchLock() {
+    void releaseFetchLock(Uuid fetchId) {
         // Register the metric for the duration the fetch lock was held. Do not register the metric
         // if the fetch lock was not acquired.
-        if (fetchLock.get()) {
-            long currentTime = time.hiResClockMs();
+        long currentTime = time.hiResClockMs();
+        if (fetchLock.compareAndSet(fetchId, null)) {
```

Comment:
Also add a non-null check in the method for `fetchId`, `Objects.requireNonNull(fetchId)`.